src/hotspot/share/runtime/arguments.cpp

```diff
@@ -838,72 +905,53 @@ static bool set_fp_numeric_flag(JVMFlag* flag, char* value, JVMFlagOrigin origin
   return false;
 }
 
-static bool set_numeric_flag(JVMFlag* flag, char* value, JVMFlagOrigin origin) {
-  julong v;
-  int int_v;
-  intx intx_v;
-  bool is_neg = false;
-
+static JVMFlag::Error set_numeric_flag(JVMFlag* flag, char* value, JVMFlagOrigin origin) {
   if (flag == NULL) {
-    return false;
+    return JVMFlag::INVALID_FLAG;
   }
 
-  // Check the sign first since atojulong() parses only unsigned values.
-  if (*value == '-') {
-    if (!flag->is_intx() && !flag->is_int()) {
-      return false;
-    }
-    value++;
-    is_neg = true;
-  }
-  if (!Arguments::atojulong(value, &v)) {
-    return false;
-  }
   if (flag->is_int()) {
-    int_v = (int) v;
-    if (is_neg) {
-      int_v = -int_v;
+    int v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_int(flag, &v, origin);
     }
-    if ((!is_neg && v > max_jint) || (is_neg && -(intx)v < min_jint)) {
-      return false;
-    }
-    return JVMFlagAccess::set_int(flag, &int_v, origin) == JVMFlag::SUCCESS;
   } else if (flag->is_uint()) {
-    if (v > max_juint) {
-      return false;
+    uint v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_uint(flag, &v, origin);
     }
-    uint uint_v = (uint) v;
-    return JVMFlagAccess::set_uint(flag, &uint_v, origin) == JVMFlag::SUCCESS;
   } else if (flag->is_intx()) {
-    intx_v = (intx) v;
-    if (is_neg) {
-      if (intx_v != min_intx) {
-        intx_v = - intx_v;
-        if (intx_v > 0) {
-          return false; // underflow
-        }
-      }
-    } else {
-      if (intx_v < 0) {
-        return false; // overflow
-      }
+    intx v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_intx(flag, &v, origin);
     }
-    return JVMFlagAccess::set_intx(flag, &intx_v, origin) == JVMFlag::SUCCESS;
   } else if (flag->is_uintx()) {
-    uintx uintx_v = (uintx) v;
-    return JVMFlagAccess::set_uintx(flag, &uintx_v, origin) == JVMFlag::SUCCESS;
+    uintx v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_uintx(flag, &v, origin);
+    }
   } else if (flag->is_uint64_t()) {
-    uint64_t uint64_t_v = (uint64_t) v;
-    return JVMFlagAccess::set_uint64_t(flag, &uint64_t_v, origin) == JVMFlag::SUCCESS;
+    uint64_t v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_uint64_t(flag, &v, origin);
+    }
   } else if (flag->is_size_t()) {
-    size_t size_t_v = (size_t) v;
-    return JVMFlagAccess::set_size_t(flag, &size_t_v, origin) == JVMFlag::SUCCESS;
+    size_t v;
+    if (parse_integer(value, &v)) {
+      return JVMFlagAccess::set_size_t(flag, &v, origin);
+    }
   } else if (flag->is_double()) {
-    double double_v = (double) v;
-    return JVMFlagAccess::set_double(flag, &double_v, origin) == JVMFlag::SUCCESS;
-  } else {
-    return false;
+    jlong v;
+    if (parse_integer(value, &v)) {
```

Comment:
I don't see how this can parse floating-point values ???

Reply:
Floating point strings with a decimal point character are parsed elsewhere in `set_fp_numeric_flag()`. I'll add a comment to clarify this.